src/components/PinInput/src/PinInputControl.vue

```diff
@@ -76,131 +90,113 @@ export default {
-			}
-			// Only allow integers in input
-			const inputValue = Number.isInteger(Number.parseInt(input, BASE_TEN)) ? input : '';
-
-			// One-time-code autofill is treated as an input, not a paste
-			if (this.attemptSplitPinIntoInputs(inputValue, index)) {
-				return;
-			}
-
-			this.$set(this.pin, index, inputValue);
+			this.handlePinChange(input);
+		},
 
-			const firstIncompleteCellIndex = this.findFirstIncompletePinCellIndex();
+		onPastePin(event) {
+			event.preventDefault();
+			const pastedValue = (event.clipboardData || window.clipboardData).getData('text');
+			this.handlePinChange(pastedValue);
+		},
 
-			// eslint-disable-next-line no-magic-numbers
-			if (firstIncompleteCellIndex === -1) {
-				this.handleComplete(this.currentPin);
+		handlePinChange(input) {
+			this.updateCaretPosition();
+			clearTimeout(this.completeTimeout);
+			if (!input) {
+				this.inputValue = input ?? '';
 				return;
 			}
 
-			// This allows us to auto jump to the next code cell input as the user types.
-			this.focusOnPinCell(firstIncompleteCellIndex);
-		},
+			this.inputValue = this.sanitizePinValue(input);
 
-		attemptSplitPinIntoInputs(value, inputIndex) {
-			if (inputIndex === 0
-				&& value?.length === this.pinLength
-				&& Number.isInteger(Number.parseInt(value, BASE_TEN))
-			) {
-				this.$set(this, 'pin', value.split(''));
-
-				// Having a timeout here gives the user a chance to see their code before success/failure
-				const TIMEOUT_LENGTH_MS = 500;
-				setTimeout(() => { this.handleComplete(this.currentPin); }, TIMEOUT_LENGTH_MS);
-				return true;
+			if (this.inputValue.length === this.pinLength) {
+				this.handleComplete(this.inputValue);
 			}
-			return false;
 		},
 
-		handleComplete(pin) {
-			this.$emit('complete', pin);
-		},
-
-		onPastePin(event, index) {
-			event.preventDefault();
-			const pastedValue = (event.clipboardData || window.clipboardData).getData('text');
-			this.attemptSplitPinIntoInputs(pastedValue, index);
+		sanitizePinValue(input = '') {
+			return input
+				.replaceAll(INPUT_FILTER_REGEX, '')
+				.slice(0, this.pinLength);
 		},
 
-		onDelete(event, index) {
-			// Reset current index, and then focus on previous cell
-			this.$set(this.pin, index, '');
-
-			// eslint-disable-next-line no-magic-numbers
-			const newIndex = index - 1;
-			if (newIndex >= 0) {
-				this.focusOnPinCell(newIndex);
+		sanitizePinInput(event) {
+			if (!event.key?.match(INPUT_KEY_REGEX)) {
+				event.preventDefault();
+				event.stopPropagation();
 			}
 		},
 
-		onFocusPin(event, index) {
-			event.preventDefault();
-			this.$set(this.pin, index, '');
+		updateCaretPosition() {
+			this.caretPosition = this.isFocused
+				? this.$refs.input.selectionStart
+				: undefined;
 		},
 
-		resetPin() {
-			const newArray = (new Array(this.pinLength)).fill('');
-			this.$set(this, 'pin', newArray);
+		handleComplete(pin) {
+			this.completeTimeout = setTimeout(() => this.$emit('complete', pin), COMPLETE_TIMEOUT_MS);
```

Comment:
Can you explain, or add a code comment, why this timeout is necessary? What goes wrong or what breaks if we just emit `complete` in the same frame that the last number is entered?

Reply:
I was trying to match the old functionality that had it on paste, but I'll rework things a bit. I realize I didn't actually hook things up for pasting.

Reply:
Removed this. The only thing is if the validation is quick, users won't see the last input value. Not sure if that's what we want.

https://user-images.githubusercontent.com/3402466/194622690-107d19a8-61b5-4568-abbd-c7d6279bdf47.mov

Reply:
i think it's okay because the invalid state / shake animation is managed by the user code not the library code, so if users want to introduce a small delay/pause before marking the pininput invalid they can do so, and the library remains unopinionated in this regard, so i think it's fine/good to keep it like this.